sound/soc/sof/ipc4-topology.c

```diff
@@ -206,6 +206,20 @@ static int sof_ipc4_setup_process_set_cdata(struct snd_sof_widget *swidget)
 		process->cdata = control_data;
 		break;
 	}
+
+	if (process->payload_with_base_cfg_ext && process->cdata) {
+		struct snd_soc_component *scomp = swidget->scomp;
+		struct sof_ipc4_base_module_cfg_ext *base_ext;
+		struct sof_ipc4_control_data *control_data;
+
+		control_data = process->cdata;
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
```

Comment:
`control_data->data` is the blob with the ABI header, `control_data->data->data` is the (supposed to be) anonymous data of the blob which is wrapped into the ABI header for validation purpose.

I don't think it is a good idea to interpret a binary blob as `sof_ipc4_base_module_cfg_ext` - which is not even an ABI struct - and feel safe about it.

The thing I'm worried is that we expect a blob to contain data which is part of the IPC messaging between host and the firmware. I would expect that the kernel would craft the base_cfg_ext alongside of the base_cfg struct when it is needed by the module, receiving the MOD_INIT_INSTANCE message, it should not belong to the blob.
This makes the blob as a sub-mini-covert topology by itself for such a modules.

Reply:
If the kernel crafts the base_cfg_ext, we should have a place that provides pin fmt for all the sink/source pins. I don't think we have except for the kcontrol.

A module should have all its kcontrols in `swidget->widget->kcontorl_news[swidget->wdiget->num_kcontrols]`, and in this PR, I only cache the kontrol for init in process->cdata. so process->cdata should have the blob for init.

we only cache the kcontrol for init  because it is unique, and it needs extra care (concatenating with base config...) .  For kcontrols used by module config/large_config, firstly, there could be more than one kcontrols, each for a param id, and second, it doesn't need special care, it is good to make the kcontrol data as payload directly.

Reply:
> If the kernel crafts the base_cfg_ext, we should have a place that provides pin fmt for all the sink/source pins. I don't think we have except for the kcontrol.

How do we configure the 'other' pins on copier for example, or other modules which does not support the base_cfg_ext via their MOD_INIT_INSTANCE message?
 
> A module should have all its kcontrols in `swidget->widget->kcontorl_news[swidget->wdiget->num_kcontrols]`, and in this PR, I only cache the kontrol for init in process->cdata. so process->cdata should have the blob for init.

That's fine, but then: if you have multiple blobs (aka bytes kcontrols) for a module, one of them is to be used via MOD_INIT_INSTANCE (MII). Should this MII blob be user update-able as well?
If the blob is big for MII then we have two places where it is stored, one in the kcontrol and one in the IPC4 mesage payload section, should they be merged and point to the same location?
 
> we only cache the kcontrol for init because it is unique, and it needs extra care (concatenating with base config...) . For kcontrols used by module config/large_config, firstly, there could be more than one kcontrols, each for a param id, and secondly, it doesn't need special care, it is good to make the kcontrol data as payload directly.

Right, so via MII we can only pass one blob (targeting a specific param_id)?
If we can update the same param_id parameter runtime via LARGE_CONFIG/MOD_CONFIG message then that blob also contains base_cfg_ext ?

I still have a feeling that the blob must not contain IPC4 message structure (which is not in ABI header). Bundling pin configuration within the blob makes a blob to specific to a certain use case while the blob itself is the parameter set for a certain function of the module, but it is not a configuration of the overall module (it's pins). It's pins should be configured based on the surroundings and connection. If the module have a single configuration for the pins and it is hardwired then why it needs to be sent to the firmware int he first place?